File: frontend/static/js/scenario.js
/*
    Copyright 2022 Jerome DE LUCCHI

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

import Script from './script.js'
import Account from './account.js'
import Node from './node.js'
import FrontendConfig from './frontend.js'

var config = new FrontendConfig()
var script = new Script()
var account = new Account()
var node = new Node()

let Scenario = class {

    constructor() {
        // all related to management run
        this._id
        this.scenarioName
        this.scenarioDescription
        this.scenarioScripts = []
        this.scenarioNodes = []
        this.scenarioAccountEmail
        this.scenarioParallelMode
        this.scenarioScriptThreads
        this._rd

    }

    set id(id) { this._id = id }
    set name(name) { this.scenarioName = name }
    set description(description) { this.scenarioDescription = description }
    set scripts(scripts) { this.scenarioScripts = scripts }
    set nodes(nodes) { this.scenarioNodes = nodes }
    set flag_parallel_mode(fpm) { this.scenarioParallelMode = fpm }
    set account_email(email) { this.scenarioAccountEmail = email }
    set max_parallel_script_threads(sst) { this.scenarioScriptThreads = sst }
    set rawData(rd) { this._rd = rd }

    get id() { return this._id }
    get name() { return this.scenarioName }
    get description() { return this.scenarioDescription }
    get scripts() { return this.scenarioScripts }
    get nodes() { return this.scenarioNodes }
    get flag_parallel_mode() { return this.scenarioParallelMode }
    get account_email() { return this.scenarioAccountEmail }
    get max_parallel_script_threads() { return this.scenarioScriptThreads }
    get rawData() { return this._rd }

    add = async (scenario) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios'
        let header = { 'Authorization': config.session.httpToken, 'Content-Type': "application/json; charset=utf-8"}
        let data = JSON.stringify({ "scenario": scenario })
        let response = await fetch(url, {method: 'POST', headers: header, body: data})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    load = (scenarioData) => {
        this.id = scenarioData["_id"]
        this.name = scenarioData["_source"]["name"]
        this.description = scenarioData["_source"]["description"]
        this.scripts = scenarioData["_source"]["scripts"]
        this.nodes = scenarioData["_source"]["nodes"]
        this.realm = scenarioData["_source"]["realm"]
        this.account_email = scenarioData["_source"]["account_email"]
        this.flag_parallel_mode = scenarioData["_source"]["flag_parallel_mode"]
        this.max_parallel_script_threads = scenarioData["_source"]["max_parallel_script_threads"]
        this.rawData = scenarioData["_source"]
    }

    list = async () => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios'
        let header = { 'Authorization': config.session.httpToken}
        let response = await fetch(url, {method: 'GET', headers: header})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    listById = async (scenarioId) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios/' + scenarioId
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: 'GET', headers: header})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    listByScriptRunId = async (scriptRunId) => {
        let url = config.proxyAPI + '/realm/' + config.session.realm + '/management/scriptrun/' + scriptRunId
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: 'GET', headers: header})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    searchByName = async (scenarioName) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios/names/' + scenarioName + '/search'
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: 'GET', headers: header})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    execute = async (scenario) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios/execute'
        let header = { 'Authorization': config.session.httpToken, 'Content-Type': "application/json; charset=utf-8" }
        let data = JSON.stringify(scenario)
        let response = await fetch(url, {method: 'POST', headers: header, body: data})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    executeById = async (scenarioId) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/scenarios/execute/' + scenarioId
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: 'GET', headers: header})
        if (response.ok) {
            response = await response.text()
            response = JSON.parse(response)
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    setFormData = async () => {
        this.formData = {
            "name": $("#scenarioName").val(),
            "description": $("#scenarioDescription").val(),
            "flag_parallel_mode": $("#scenarioParallelMode").is(":checked") ? true : false,
            "max_parallel_thread": $("#maxParallelThread").val()
        }
    }

    rolesSelectedNodes = () => {
        let roles = []
        node.listByNames(this.scenarioNodes).then(function(nodesData) {
            nodesData["hits"]["hits"].forEach(function(nodeSource) {
                nodeSource["_source"]["role"].forEach(function(role) {
                    if ( Object.keys(role).includes("application") ) {
                        if ( ! roles.includes(role["application"])) {
                            roles.push(role["application"])
                        }
                    } else if ( Object.keys(role).includes("protocol")) {
                        if ( ! roles.includes(role["protocol"])) {
                            roles.push(role["protocol"])
                        }
                    }
                })
            })
        })
    }
}

export default Scenario

File: backend/api/scenario.py
# -*- coding:utf-8 -*-
"""
   Copyright 2022 Jerome DE LUCCHI
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
       http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

import re
import json
import threading
import base64
import string
import random
from api import statistic
from api import scenarioManager
from api.cluster import Cluster
from api.infra import Infra
from api.node import Node


class Scenario:
    def __init__(self, connector):
        self.ES = connector
        self.DB_INDEX = 'blast_obj_scenario'
        self.STATISTIC = statistic.Statistic(self.ES)
        self.STATISTIC_DATA = self.STATISTIC.STATISTIC_DATA
        self.STATISTIC_DATA["object_type"] = 'scenario'

    def add(self, account_email: str, realm: str, scenario: dict):

        try:
            if scenario["name"] == "":
                raise Exception("Scenario name is required. Empty value is not accepted.")

            scenario_name_pattern = re.compile("[a-zA-Z0-9\-_]+")
            if not scenario_name_pattern.fullmatch(scenario["name"]):
                raise Exception("Cluster name is not valid. Alphanumeric characters and '_', '-', are accepted.")

            if len(scenario["nodes"]) == 0:
                raise Exception("Scenario nodes must not be null. Select one node at least.")

            if len(scenario["scripts"]) == 0:
                raise Exception("Scenario scripts must not be null. Select one script at least.")

            scenario["realm"] = realm
            scenario["account_email"] = account_email
            scenario_add_res = self.ES.index(index=self.DB_INDEX, body=json.dumps(scenario), refresh=True)
            if scenario_add_res["result"] != "created":
                raise Exception("Internal Error: Scenario create failure.")

            self.STATISTIC_DATA["object_action"] = 'create'
            self.STATISTIC_DATA["account_email"] = account_email
            self.STATISTIC_DATA["realm"] = realm
            self.STATISTIC_DATA["timestamp"] = statistic.UTC_time()
            self.STATISTIC_DATA["object_name"] = scenario["name"]
            self.STATISTIC.add(self.STATISTIC_DATA)

            return scenario_add_res

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:__add__")
            print(e)
            return {"failure": str(e)}

    def delete(self, realm: str, scenario_ids: list):

        try:
            req = json.dumps(
                {
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "term": {
                                        "realm": realm
                                    }
                                },
                                {
                                    "terms": {
                                        "_id": scenario_ids
                                    }
                                }
                            ]
                        }
                    }
                }
            )
            return self.ES.delete_by_request(index=self.DB_INDEX, body=req, refresh=True)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:__delete__")
            print(e)
            return {"failure": str(e)}

    def execute(self, scenario=None, scenario_realm=None, scenario_id=None):
        """
        This function execute a scenario order coming from the API

        [1]: the scenario is executed from the GUI:
        + the scenario_realm is collected from the API URL
        + the scenario_id is empty
        + the scenario is the details of the items selected by the operator

        :param scenario: A copy of object to be executed
        :param scenario_id: Id of the object to be executed
        """
        try:
            if scenario["name"] == "":
                raise Exception("Scenario name is required to execute.")

            if len(scenario["nodes"]) == 0:
                raise Exception("Scenario nodes is required to execute. Select one node at least.")

            if len(scenario["scripts"]) == 0:
                raise Exception("Scenario scripts is required to execute. Select one script at least.")

            scenario_manager = scenarioManager.ScenarioManager(self.ES)
            scenario_id = str("oneshot-" + base64.urlsafe_b64encode(''.join([random.choice(string.ascii_letters + string.digits) for n in range(16)]).encode('utf-8')).decode("utf-8"))
            execute_scenario_kwargs = {
                "scenario_realm": scenario_realm,
                "scenario": scenario,
                "scenario_id": scenario_id,
                "execution_id": str(base64.urlsafe_b64encode(''.join([random.choice(string.ascii_letters + string.digits) for n in range(16)]).encode('utf-8')).decode("utf-8"))
            }

            new_t = threading.Thread(target=scenario_manager.execute_scenario, args=(), kwargs=execute_scenario_kwargs)
            new_t.start()
            return {"success": "started"}

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:api_execute")
            print(e)
            return {"failure": str(e)}

    def list(self, realm: str):

        try:
            req = json.dumps(
                {
                    "size": 10000,
                    "query": {
                        "term": {
                            "realm": realm
                        }
                    },
                    "sort": [
                        {
                            "name": {
                                "order": "desc"
                            }
                        }
                    ]
                }
            )
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:__list__")
            print(e)
            return {"failure": str(e)}

    def list_by_script_id(self, realm: str, script_id: str):
        """ Returns all the scenario saved with the given script id """
        try:
            req = json.dumps(
                {
                    "size": 10000,
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "term": {
                                        "realm": realm
                                    },
                                    "scripts": {
                                        "term": script_id
                                    }
                                }
                            ]
                        }
                    }
                }
            )
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:list_by_script_id")
            print(e)
            return {"failure": str(e)}

    def list_any_by_script_id(self, script_id: str):
        """ Returns all the scenario saved with the given script id """
        try:
            req = json.dumps(
                {
                    "size": 10000,
                    "query": {
                        "bool": {
                            "filter": {
                                "scripts": {
                                    "term": script_id
                                }
                            }
                        }
                    }
                }
            )
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:list_any_by_script_id")
            print(e)
            return {"failure": str(e)}

    def list_by_node_id(self, realm: str, node_id: str):
        """ Returns all the scenario saved with the given node id """
        try:
            req = json.dumps(
                {
                    "size": 10000,
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "term": {
                                        "nodes": node_id
                                    }
                                },
                                {
                                    "term": {
                                        "realm": realm
                                    }
                                }
                            ]
                        }
                    }
                }
            )
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:list_by_node_id")
            print(e)
            return {"failure": str(e)}

    def list_by_id(self, realm: str, scenario_id: str):

        try:
            req = json.dumps(
                {
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "term": {
                                        "realm": realm
                                    }
                                },
                                {
                                    "terms": {
                                        "_id": scenario_id
                                    }
                                }
                            ]
                        }
                    }
                }
            )
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print("backend Exception, file:scenario:class:scenario:func:list_by_ids")
            print(e)
            return {"failure": str(e)}

    def map_id_name(self, realm: str, scenario_id: str):

        try:
            return self.list_by_id(realm, scenario_id)["hits"]["hits"][0]["_source"]["name"]

        except Exception as e:
            print(e)
            return {"failure": str(e)}

    def update(self, scenario_id: str, scenario: dict):
        pass
